File: backend/services/generateAccessToken.js
// Take the Refresh Token from env and
// generate a new access token

import {setToken} from '../store/tokenStore.js';

export const generateAccessToken = async () => {
  const clientId = process.env.ZOHO_ANALYTICS_CLIENT_ID;
  const clientSecret = process.env.ZOHO_ANALYTICS_CLIENT_SECRET;
  const refreshToken = process.env.ZOHO_ANALYTICS_REFRESH_TOKEN;
  const baseUrl = process.env.ZOHO_ACCOUNT_BASE_URL;

  if (!clientId || !clientSecret || !refreshToken || !baseUrl) return null;

  const params = new URLSearchParams({
    refresh_token: refreshToken,
    client_id: clientId,
    client_secret: clientSecret,
    grant_type: 'refresh_token',
  });

  const url = `${baseUrl}/oauth/v2/token?${params.toString()}`;

  try {
    const response = await fetch(url, {method: 'POST'});
    const data = await response.json();

    if (!response.ok) return null;

    setToken(data.access_token, data.expires_in);

    return data.access_token;
  } catch (error) {
    console.error('Refresh token error:', error);
    return null;
  }
};

File: src/main/java/com/anselumjuju/services/accessToken/GenerateAccessToken.java
package com.anselumjuju.services.accessToken;

import com.anselumjuju.lib.EnvConfig;
import com.anselumjuju.stores.TokenStore;
import com.anselumjuju.utils.Utils;
import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.util.Map;

public class GenerateAccessToken {
    public static String generateAccessToken() {
        String refreshToken = EnvConfig.ZOHO_ANALYTICS_REFRESH_TOKEN;
        String clientId = EnvConfig.ZOHO_ANALYTICS_CLIENT_ID;
        String clientSecret = EnvConfig.ZOHO_ANALYTICS_CLIENT_SECRET;
        String authUrl = EnvConfig.ZOHO_ACCOUNT_BASE_URL;

        String params = String.format(
                "refresh_token=%s&client_id=%s&client_secret=%s&grant_type=refresh_token",
                Utils.encode(refreshToken),
                Utils.encode(clientId),
                Utils.encode(clientSecret)
        );

        String url = authUrl + "/oauth/v2/token?" + params;

        try (HttpClient client = HttpClient.newHttpClient()) {
            HttpRequest request = HttpRequest.newBuilder()
                    .uri(new URI(url))
                    .POST(HttpRequest.BodyPublishers.ofString(params))
                    .build();

            HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());

            Gson gson = new Gson();
            Map<String, Object> body = gson.fromJson(
                    response.body(),
                    new TypeToken<Map<String, Object>>() {
                    }.getType()
            );

            String accessToken = String.valueOf(body.get("access_token"));

            String expiryIn = String.valueOf(body.get("expires_in"));
            String expiryTime = String.valueOf(System.currentTimeMillis() + ((long) Double.parseDouble(expiryIn) * 1000));

            TokenStore.setAccessToken(accessToken);
            TokenStore.setExpiryTime(expiryTime);

            return accessToken;
        } catch (Exception e) {
            System.out.println("Error generating access token " + e.getMessage());
            return "";
        }
    }
}
